server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -357,6 +364,26 @@ static int shortcutTotalHitCount(IndexReader reader, Query query) throws IOExcep
                 count += context.reader().docFreq(term);
             }
             return count;
+        } else if (query.getClass() == DocValuesFieldExistsQuery.class && reader.hasDeletions() == false) {
+            final String field = ((DocValuesFieldExistsQuery) query).getField();
+            int count = 0;
+            for (LeafReaderContext context : reader.leaves()) {
+                FieldInfos fieldInfos = context.reader().getFieldInfos();
+                FieldInfo fieldInfo = fieldInfos.fieldInfo(field);
+                if (fieldInfo != null) {
+                    if (fieldInfo.getPointIndexDimensionCount() > 0) {
+                        count += PointValues.getDocCount(context.reader(), field);
```

Comment:
This method is supposed to be called on top-level readers, as reader#leaves is expected to fail on sub readers (see LeafReaderContext#leaves). Can we do `points = context.reader().getPointValues(field); if (points != null) count += points.getDocCount();` like for terms?